internal/vm/setup.go

```diff
@@ -76,6 +76,28 @@ _G["process_auto_fields"] = function(plugin)
   return json.encode(p)
 end
 
+_G["process_auto_fields_and_transform"] = function(plugin)
+  if type(plugin) ~= "string" then
+    return nil, "invalidate type, expected string"
+  end
+  local tbl, err = json.decode(plugin)
+  if err then
+    return nil, err
+  end
+  local p = Plugins:process_auto_fields(tbl, "insert")
+  -- config is empty, remove it from the plugin to avoid bad json array encoding
+  if p["config"] ~= nil and next(p["config"]) == nil then
+    p["config"] = nil
+  end
```

Comment:
style: `p["config"]` is exactly the same as `p.config` (compiles to the same bytecode).  some static checkers find it easier to follow. (since it signifies a "struct" kind of table, not an open ended one)

paranoia: calling `next()` with a non-table value results in a hard error.  it's ok to check the type of a field to verify its existence, since if it's missing, it will be the same as checking the type of `nil`, which is `"nil"`

```suggestion
  if type(p.config) == "table" and next(p.config) == nil then
    p.config = nil
  end
```